src/main/scala/com/exasol/cloudetl/filesystem/FileSystemManager.scala

```diff
@@ -1,19 +1,55 @@
 package com.exasol.cloudetl.filesystem
 
+import java.io.FileNotFoundException
+import java.nio.file.Path
+
+import org.apache.hadoop.fs.{Path => HadoopPath}
+import org.apache.hadoop.fs.FileStatus
 import org.apache.hadoop.fs.FileSystem
-import org.apache.hadoop.fs.Path
 
-object FileSystemManager {
+/**
+ * A filesystem helper class.
+ */
+final case class FileSystemManager(fileSystem: FileSystem) {
 
-  def globWithLocal(path: java.nio.file.Path, fs: FileSystem): Seq[Path] =
-    globWithPattern(path.toAbsolutePath.toUri.getRawPath, fs)
+  /**
+   * Returns list of files from path with or without globbing.
+   *
+   * It also adds paths from the sub-directories to the returned list.
+   *
+   * @param path input path
+   * @return list of paths
+   */
+  def getFiles(path: String): Seq[HadoopPath] = globStatus(path)
 
-  def globWithPattern(pattern: String, fs: FileSystem): Seq[Path] =
-    glob(new Path(pattern), fs)
+  /**
+   * Returns list of files from local path.
+   */
+  def getLocalFiles(path: Path): Seq[HadoopPath] =
+    getFiles(path.toAbsolutePath().toUri().getRawPath())
 
-  def glob(path: Path, fs: FileSystem): Seq[Path] = {
-    val opt = Option(fs.globStatus(path, HiddenFilesFilter))
-    opt.fold(Seq.empty[Path])(_.toSeq.map(_.getPath))
+  private[this] def globStatus(path: String): Seq[HadoopPath] = {
+    val hadoopPath = new HadoopPath(path)
+    val statuses = fileSystem.globStatus(hadoopPath, HiddenFilesFilter)
+    if (statuses == null) {
+      throw new FileNotFoundException(
+        s"Provided file path '$path' does not exist. Please use valid path."
+      )
+    } else {
+      if (isSingleDirectory(statuses)) {
+        // User path is a single directory without anyany glob, only then list the files.
+        listFiles(fileSystem.listStatus(statuses(0).getPath(), HiddenFilesFilter))
+      } else {
+        listFiles(statuses)
+      }
+    }
   }
 
+  // Check if user provided path returns a single directory
+  private[this] def isSingleDirectory(paths: Array[FileStatus]): Boolean =
+    paths.length == 1 && fileSystem.getFileStatus(paths(0).getPath()).isDirectory()
+
+  private[this] def listFiles(paths: Array[FileStatus]): Seq[HadoopPath] =
+    paths.toSeq.filter(_.isFile()).map(_.getPath())
```

Comment:
you're still loading all file names into memory, but I think that's accepatble

Reply:
Yes, this should be okay. 

On the other hand, I cannot really stream them. The `globStatus` already returns array. The second case is when there is single directory, we run single `listStatus`. Both of these functions already return in memory array.